hnix-store-core/src/System/Nix/Nar.hs

```diff
@@ -1,279 +1,160 @@
-            in int len <> pad len t
-
-        putContents :: Int64 -> BSL.ByteString -> B.Put
-        putContents fSize bs = str "contents" <> int fSize <> (pad fSize bs)
-        -- putContents fSize bs = str "contents" <> int (BSL.length bs) <> (pad fSize bs)
-
-        int :: Integral a => a -> B.Put
-        int n = B.putInt64le $ fromIntegral n
-
-        pad :: Int64 -> BSL.ByteString -> B.Put
-        pad strSize bs = do
-          B.putLazyByteString bs
-          B.putLazyByteString (BSL.replicate (padLen strSize) 0)
-
-        strs :: [BSL.ByteString] -> B.Put
-        strs = mapM_ str
+  -- * Internal
+  , streamNarIO
+  ) where
 
+import qualified Control.Concurrent     as Concurrent
+import           Control.Monad          (when)
+import qualified Control.Monad.IO.Class as IO
+import           Data.Bool              (bool)
+import qualified Data.ByteString        as BS
+import qualified Data.ByteString.Char8  as BSC
+import qualified Data.ByteString.Lazy   as BSL
+import           Data.Foldable          (forM_)
+import qualified Data.List              as List
+import           Data.Monoid            ((<>))
+import qualified Data.Serialize.Put     as Serial
+import           GHC.Int                (Int64)
+import qualified System.Directory       as Directory
+import           System.FilePath        as FilePath
+import qualified System.IO              as IO
 
-------------------------------------------------------------------------------
--- | Deserialize a Nar from lazy ByteString
-getNar :: B.Get Nar
-getNar = fmap Nar $ header >> parens getFile
-    where
+import qualified System.Nix.Internal.Nar.Effects as Nar
+import qualified System.Nix.Internal.Nar.Parser  as Nar
 
-      header   = assertStr "nix-archive-1"
 
 
-      -- Fetch a FileSystemObject
-      getFile = getRegularFile <|> getDirectory <|> getSymLink
 
-      getRegularFile = do
-          assertStr "type"
-          assertStr "regular"
-          mExecutable <- optional $ Executable <$ (assertStr "executable"
-                                                   >> assertStr "")
-          assertStr "contents"
-          (fSize, contents) <- sizedStr
-          return $ Regular (fromMaybe NonExecutable mExecutable) fSize contents
+-- For a description of the NAR format, see Eelco's thesis
+-- https://nixos.org/%7Eeelco/pubs/phd-thesis.pdf
 
-      getDirectory = do
-          assertStr "type"
-          assertStr "directory"
-          fs <- many getEntry
-          return $ Directory (Map.fromList fs)
 
-      getSymLink = do
-          assertStr "type"
-          assertStr "symlink"
-          assertStr "target"
-          fmap (SymLink . E.decodeUtf8 . BSL.toStrict) str
+buildNarIO
+  :: Nar.NarEffects IO
+  -> FilePath
+  -> IO.Handle
+  -> IO ()
+buildNarIO effs basePath outHandle = do
+  streamNarIO (\chunk -> BS.hPut outHandle chunk >> Concurrent.threadDelay 10) effs basePath
+
+
+-- | This implementation of Nar encoding takes an arbitrary @yield@
```

Comment:
Move this implementation under `System.Nix.Internal.Nar` for uniformity with the parser